ENH,API: Store exported buffer info on the array

seberg opened the pull request:
This speeds up array deallocation and buffer exports, since it
removes the need to global dictionary lookups.  It also somewhat
simplifies the logic.  The main advantage is prossibly less the
speedup itself (which is not large compared to most things that
happen in the livetime of an array), but rather that no unnecessary
work is done for shortlived arrays, which never export a buffer.

The downside of this approach is that the ABI changes for anyone
who would be subclassing ndarray in C.

---

I am not super invested in this, but I do feel its worth considering.  This requires breaking the ABI slightly (so would need release notes), because it extends the array struct (and void, but that is much less likely to be used).  Marking as draft since there are probably some smaller moving parts.  But since I was looking at the buffer interface today, thought I would wrap this up and show it at least.

seberg committed:
ENH,API: Store exported buffer info on the array

This speeds up array deallocation and buffer exports, since it
removes the need to global dictionary lookups.  It also somewhat
simplifies the logic.  The main advantage is prossibly less the
speedup itself (which is not large compared to most things that
happen in the livetime of an array), but rather that no unnecessary
work is done for shortlived arrays, which never export a buffer.

The downside of this approach is that the ABI changes for anyone
who would be subclassing ndarray in C.

seberg commented:
The very real alternative to this would be to just implement the dealloc/free slot of the buffer interface.  That will break some weirder usage, but maybe its much less then it used to be (plus scalars may have been part of the issue before).  So maybe more a major release try (I just don't know how much it would affect).

mattip commented on numpy/core/include/numpy/ndarraytypes.h:
```diff
@@ -701,6 +701,7 @@ typedef struct tagPyArrayObject_fields {
     int flags;
```
Could we use flags to version this struct and indicate whether `_buffer_info` exists or not? If it doesn't exist we should raise an error when trying to use it.

seberg replied:
I am not sure I am following.  We could add a version, but I guess we already have the NumPy version to find out how large the struct will be (or even the type object).  If we would version it, C-side subclassers (if we can even figure out someone is that), would need to opt-in with a flag and we otherwise keep using the global? (and maybe try to give a deprecation warning)

mattip replied:
I was thinking more along the lines of "Detect that we did not set up this `PyArrayObject_fields`, refuse to do any buffer processing". That would force anyone mimicking our struct to also add the field and update the flag, otherwise they would get errors. So then there are two parts: 
- set something inside the struct itself to show it has this extra field
- detect this thing and raise an error rather than write past the end of the struct

It needs to be inside the struct since the struct can be fed in by C-API consumers.

mattip replied:
I would be happy to get rid of the global registry and force the few people who reproduce the `PyArrayObject_fields` struct to update or fail.

seberg replied:
Good point, I have to think whether its feasible/easy, since python subclasses must not notice any of such things.

seberg replied:
@mattip, I guess we could initialize the field with a singleton/garbage (instead of NULL), in the hope that it makes clashes/crashes more likely.  The only usage that would really find, is if the subclasser also initializes the field to NULL and does not end up using it consistently.  (Since any non-NULL value should lead to a crash during dealloc.)

I am not sure there is a good way to find C-side subclassing, aside from checking whether the types `basicsize` changed, which would be the case unless the subclasser added a single pointer at the end.  So that is something we can do: raise an error if the types size differs, but there isn't some additional flag set.  But it seems a bit convoluted? We would have to set that flag on each instance probably, or require reimplementation of the buffer export, which seems strange?

mattip replied:
This is the scenario that worries me:
- Someone uses the older C-API to create an ndarray, using something like `PyArray_NewFromDescr`. Now they have a `PyArrayObject` that is smaller than the one after this PR.
- They somehow pass it to a newer NumPy. Now we get a `PyObject*`, and 
- NumPy 1.20 casts it to a `PyArrayObject`. Whoops, we will read this field off the end of the memory they gave us.

This is similar to the problems we were having with Cython compiled code like cython/cython#2221 that led to cython/cython#2627.

`sizeof()` or junk memory will not help here. I think the only way to detect this is to have some kind of versioning in the struct itself, which is why I suggested we add a value to the flag, I think something like `#define NPY_ARRAY_HAS_BUFFER_INFO 0x4000` might work. Then we can set the flag whenever we allocate and initialize a new `PyArrayObject_fields`, and check that the flag has that value set whenever we wish to manipulate the `_buffer_info`.

mattip replied:
`basicsize` might work as well.

seberg replied:
So the example we want to guard against is mainly this type of thing?:
```
cimport numpy as npc

cdef class sub(npc.ndarray):
    pass
```
without any extra fields the struct will be too small (I thought cpython would check this, but it does not), if we add more fields they will be incompatible.  The case where the struct is too small is simple though: We can just raise an error in `tp_alloc`.  But if you add a single new field there, we will not know and need the flag (or hope for the best with guards).
I am actually not sure where the flag should be set. `tp_alloc` is plausible, but not sure its available in cython, I guess in cython you may have to set it in `__array_finalize__` (but only if numpy is new enough).

For heaptypes, we would have to walk the mro (to see if there is a static subclass involved) – they always have at least one additional field `tp_dict`, and may have more for `__slots__`. Or simply assume that there is no static subclass involved.

I am not sure about `PyArray_NewFromDescr`, you can't create an incompatible `ndarray` in any reasonable way I think. (It would have to be a stack allocated array, which is possible, but pretty dirty – at least outside of NumPy).

mattip replied:
Let's discuss the implications at the next triage meeting.

seberg replied:
OK, thinking about it a bit more. I don't have a big new idea.  We noted flags in the meeting, there is the option to:

1. Add a flag to indicate that a buffer was exported, the only use would be a sanity check on `dealloc` (so that we do not crash hard).
2. Somehow allow downstream to indicate that they are aware of a growing struct. But I don't see how downstream can add a flag here, since it needs to be flagged on the type object.

There is no perfect solution, since point 1 can easily crash hard if an array exports a buffer and then the field gets modified.  That is super rare, but I thought an alternative could be to tag the pointer to give us a chance to convert the hard crash into a useful message.

I was going to polish this up a bit, but it would be good to finish gh-16936 and gh-17295 since it modifies the same code.

mattip replied:
Isn't [the `PyArray_GetNDArrayCVersion` mechnism](https://numpy.org/doc/stable/reference/c-api/array.html?highlight=pyarray_getndarraycversion#c.PyArray_GetNDArrayCVersion) supposed to adequately protect us from ABI changes?

seberg replied:
Ooops, yeah. That forces everyone to recompile even though almost nobody will have to, but is maybe best. But, it seems the current check won't allow to easily compile against both versions of NumPy, which is probably still problematic (unless we assume that this happens so rarely that packaging isn't probably an issue for these projects)?

mattip replied:
Maybe try building this PR and running it against an version of SciPy compiled for 1.19 or older

mattip replied:
I think PyArray_GetNDArrayCVersion works properly when mixing NumPy versions in, say, SciPy and Numpy. Otherwise we would have heard loud screams when we modified PyUFuncObject in v1.16 and changed the ABI version number. So I am 99% sure this discussion is a non-issue. I guess one way to find out is to merge this now, and wait for the weekly builds to percolate through the system.

mattip replied:
... which means fixing the merge conflict and getting tests to pass.

seberg replied:
I hope it is, but check here:

https://github.com/numpy/numpy/blame/39c915c1236d61debb4b3e10dfadf910da6c9b30/numpy/core/setup_common.py#L24

We have the API version and the ABI version. The API version we can increase easily, since it will only affect anything if you compile against a new NumPy version and then run it on an old one (which we can assume nobody does). But the ABI version has not been touched in 10 years.  Now I don't know how well that worked out 10 years ago, but it seems like a time before widespread comfortable pip?

mattip replied:
OK, then I guess we need to do some testing to make sure.

mattip replied:
I played with this for a while with `pip install scipy` and a local `numpy` build (after changing `C_ABI_VERSION` in `numpy/core/setup_common.py`. I also modified `_import_array` in `b/numpy/core/code_generators/generate_numpy_api.py` to print when it is called. If I import `scipy`, I see it calls `_impport_array`, and the `C_ABI_VERSION` is the new one. Which makes sense, the function is exposed via `_multiarrayumath.so`, which comes from NumPy, but I just wanted to confirm.

seberg replied:
I don't understand.  I just tried as well. First, I had to clean up to get the new C-ABI version, I think.  Second `import_array` is defined as part of `__multiarray_api.h` and should be included as a header and not linked? If it is linked, that would defeat the purpose.  So I assume your pip install compiled scipy?

Importing the system ScPy after incrementing the C-ABI version, I get this (version was 100009 or so before):
```
In [2]: f"{np.core._multiarray_umath._get_ndarray_c_version():x}"                                                       
Out[2]: '200000a'

In [3]: from scipy import linalg                                                                                        
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
RuntimeError: module compiled against ABI version 0x1000009 but this version of numpy is 0x200000a
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
ImportError: numpy.core.multiarray failed to import
```

mattip replied:
OK, I didn't try importing scipy.linalg (which I guess is compiled by cython). Thanks for checking more deeply. So we cannot change C_ABI_VERSION without recompiling the world.

Where does that leave us with this PR (and the one to configure memory allocations)?

seberg replied:
I guess we either:

1. Wait until some day where we actually do a big bang NumPy 2.0 (I am fan of a 2.0, but not of recompile the world 2.0)
2. Define the size of these structs as "undefined"? Placing it into the downstream responsibility of checking for binary compatibility in this case (this is what cython normally does, although with a warning? I have yet to test it here though).
3. We think about some extremely slow route, with compile time warnings...?

My opinion about option 2 has not changed (with two caveats):

* I would like to give a reasonable error/warning at least in the majority of (hypothetical) cases. This is to *decrease the severity* if the issue appears.
* No big "user" of this shows up and shouts loudly. Which is, unfortunately, plausible.

Of course, if you go by strict policies, you might just dismiss my opinion. But IMO, we should be using quality assurance type triaging, which means map out how likely it is (how many users are affected), how sever it is:

* Affects: Exceedingly few users (we are not aware of any, none of the large open source pages seem affected)
* Severity: Crash bug, which is easy to fix by recompiling once the issue is found.  That is fairly sever, but super high, since it is seems much impossible that you get wrong results without a loud crash as well.

So because it should be reasonable to fix, and be a loud crash that affects very few users. My personal opinion is that it might well be much lower impact than some "bug fixes".

Of course, you have to weigh that with the benefit. In this case probably a few percent of speedup for many scripts relying on small to mid-sized arrays and simpler code.  And later your change, which might make some big-data  use cases much faster.

seberg replied:
On the up-side here, I was trying around with cython code such as:
```
cdef class myarr(np.ndarray):
    cdef int myint  # or just pass
```
And even that crashes hard. I guess because cython supports cyclic GC, but we do not.  So that rules out one set of potential users who run into this unwittingly...

I wonder if could declare `np.ndarray` as final in the `__init__.pyd`. Or in some other way telling cython that `cdef` style subclassing is *not* valid (or at least not simple).

mattip commented:
In general I am +1 on moving the information to be local.

mattip commented:
In the last triage discussion we decided to push ahead with adding a field to the struct. It is still marked "draft" and has a merge conflict.

seberg commented:
I didn't fix the merge conflict yet because I was hoping to do gh-17295 first.

seberg commented:
@mattip OK, just marking as ready, here are the things we need to consider:

1. I currently tag the pointers (by adding 2), the reason for this is that it gives a good probability of raising an error (at least often). Instead of just crashing hard (or maybe missing the incompatibility).
2. I also tag the pointer for the `void` scalar inside `tp_alloc`, which feels a bit weird. We already allocate the store space in `gentype_alloc` which is probably just as dirty, though.  I am pretty certain now that this is OK, but maybe worth a quick thought by someone else.
3. As noted in the release note, the definition of `NPY_SIZEOF_PYARRAYOBJECT` is now not a constant anymore, the only project I have yet found that might be hit by this is here: https://github.com/patmiller/bignumpy/blob/master/bignumpy.c (opened an issue there, but it seems inactive). That project would not compile successfully on a new NumPy and would require (fairly simple adaptation). It could make sense to explain how to do it, but I am still a bit unsure if we actually want anyone subclassing from C...

@scoder as a test, I tried to do the dirty thing and subclass NumPy in cython using:
```
cimport numpy as np

cdef class MyArr(npc.ndarray):
    pass
``` 
This goes down in flames, checking `gc.is_tracked(MyArr(3))` shows that cython sets the garbage collection flag and adds it to be tracked, I assume that causes the failure in the end, but I am not sure.  It probably doesn't really matter, it would mainly be interesting if there was a simple way to make this work that I am missing.
(Overall, cython users should be happy with this, since it shaves off a bit of overhead from cython )

seberg commented:
Hmm, PyPy crashes, I guess the `tp_alloc` is likely just wrong, could just stop tagging the NULL pointer (or remove all the tagging), that way initializing all to 0 (which is what `tp_alloc` should do) is definitely the right thing.

seberg committed:
MAINT: Do not tag the NULL (no buffers exported)

The allocation is not the right place to initialize to anything but
NULL, so take the easy path and do not tag the NULL default.

mattip commented:
@seberg is this ready for final review?

seberg commented:
Its ready for review, I will add see if I can add a test for the "error" message (only the buffer export path).

seberg committed:
TST: Add test for best try RuntimeError on corrupt buffer-info

seberg commented:
@mattip the only real remaining issue that I can think of is whether we have to do something about the fact that `NPY_SIZEOF_ARRAYOBJECT` is not a constant anymore (documentation, or otherwise).

mattip commented on doc/release/upcoming_changes/16938.c_api.rst:
```diff
@@ -0,0 +1,17 @@
+
+    #define NPY_SIZEOF_PYARRAYOBJECT (PyArray_Type.tp_basicsize)
+
+since the size should not be considered a compile time constant.
```
```suggestion
Making this a compile-time constant may lead to bugs, but it is part of the public headers.
```

seberg replied:
How about:

> This was previously a compile-time constant, which could lead to bugs when running in a different NumPy version.

One thing we could do is adding a new `#define NPY_SIZEOF_PYARRAYOBJECT_STATIC sizeof(PyArrayObject_fields)` so that it is easier to replace the safe version with the unsafe one where necessary (it would be the users job to check this on import time).  But as I don't know anyone actually using it, I am somewhat OK with forcing users into `sizeof(PyArrayObject_fields)` using the deprecated API.

mattip replied:
I think any attempt at using this is doomed to failure. Can we deprecate the macro, with a message that it is the wrong thing to checke?

mattip replied:
I think we must disallow its use. These are exactly the use cases we want to loudly fail with the new code since they are depending on something staying the same size when it doesn't.

seberg replied:
@mattip I agree.  Did you formed a clear opinion on what to do here?  No `NPY_SIZEOF_PYARRAYOBJECT_STATIC` sounds good to me.  But I need to make a call what to do specifically for this PR.

* Additionally to making it a run-time constant (if that even matters for anyone), we add a deprecated tag in C?
* We just remove it (i.e. make it an `# error`)... The `(PyArray_Type.tp_basicsize)` doesn't have much point probably?
* Only deprecate it, which makes it easier to compile code that is not future proof without realizing it.

I am OK with both of the first versions, the `# error` might be nice, since it allows us to write a few words about what is going on in a very clear way that can't get lost.

mattip replied:
we could try to do some [static assert](https://stackoverflow.com/questions/3385515/static-assert-in-c) complicated thing, or just remove it and document the removal in the release notes. I think removing it is the easiest and will get the desired response if anyone is actually using it.

seberg replied:
OK, the last commit just deletes this `#define` and adds some info to the `PyArrayObject` structure documentation.

Not sure where we stand with putting this in (or more when), seems like not for 1.20 most likely, so will have to update the text slightly again before merging finally.

charris commented:
Ping all.

eric-wieser commented:
With this change, we're essentially saying users must either:

* Not extend `ndarray` in C at all
* Rebuild their C extensions before / after this change
* Do some unpleasant pointer tricks to deal with a runtime offset of the fields they're interested in (and fill tp_basicsize with a computer value too)

mattip commented:
@eric-wieser yes, that is what this PR does. We have been trying to hide `PyArrayObject_fields` since version 1.7. Let's try asking on the mailing list if there are projects this would cause problems for.

charris commented:
Sounds too risky for 1.20.

seberg committed:
Remove NPY_SIZEOF_PYARRAYOBJECT and add some documentation

charris commented:
I'm leaning to putting this in for 1.20.0, but you may want to keep your branch in the unlikely case we need to revert it.

seberg committed:
Use 3 to tag the pointer and object for a "bad" one in the test

seberg commented:
Just a minor change, using 3 now to tag (not sure that is best) and tweaked the test.  There was nothing wrong with the code, but it I tested with a function pointer, and it seem they alignment can be only 2...

charris commented:
Why is the size of the structure not a compile time constant?

seberg commented:
Hmm, maybe not quite the right wording? The size of the struct depends on the NumPy version, so it is a NumPy compile time constant, but for third party modules it is only a runtime constant.

charris commented on numpy/core/include/numpy/ndarraytypes.h:
```diff
@@ -720,7 +721,6 @@ typedef struct tagPyArrayObject {
 #endif
 
-#define NPY_SIZEOF_PYARRAYOBJECT (sizeof(PyArrayObject_fields))
 
```
What was wrong with this?

seberg replied:
Nothing as such, but it is a trap, and most code that uses it, such as:
```
struct MySubclass {
    char space_for_ndarray[NPY_SIZEOF_PYARRAYOBJECT];
    int myslot;
}
```
is probably problematic and most likely not be binary compatible with all numpy versions.  So it gives a false sense security?

charris replied:
I don't see why, it is defined at compile time just as before. The API is changed anyway, so apps compiled against 1.20 are not guaranteed to be backwards compatible.

charris replied:
And code that extends NumPy internals has never been guaranteed to be forward compatible.

seberg replied:
True, this is only interesting for forward compatibility with possible future size changes.  I would have a faint hope that it may help someone realize that since their module doesn't compile on 1.20+, it has to be modified even if you compile it with an earlier version to run in 1.20+...

I agree, we probably never promised not to modify struct sizes in this way, we have certainly done with many structs (aside from `ndarray` itself).

The first version here, I replaced the macro with `PyArray_Type.tp_basicsize` to make it a runtime constan, which is safe and will still fail if used for static sizes. (It is also probably a pretty useless macro, but...)

charris replied:
I suppose the argument is to break the compilation of downstream apps that extend the structure, but if they are compiling against 1.20 they cannot expect to be backwards compatible in any case, and their solution would be the same workaround you use here.

charris replied:
> but it cannot do so at compile time.

Could you explain that? `sizeof` is a compiler operator.

> Sizeof is a much used operator in the C or C++. It is a compile time unary operator which can be used to compute the size of its operand.

seberg replied:
But someone who uses this macro currently will incorrectly believe that their code is forward compatible. They are even avoiding the "deprecated API" to ensure binary compatibility with a potential NumPy release messing with the structs internals.

charris replied:
Why not just rename it then? Put an underscore on it or something.

charris replied:
At least just comment it out with an explanation of why it is done. The first place people will look for an explanation of trouble will be in the include file.

seberg replied:
OK, let me make this macro a static assert as Matti posted.  Even if a compiler should not support it the compiler error will point to some text regarding to what is going on...

seberg replied:
Hmm, although, I am not sure that actually compiles well always (i.e. without super confusing things). The stack-overflow posts seem to a be a bit unsure about it. Some people seem to suggest the weird:
```
static int assert_NPY_SIZEOF_PYARRAYOBJECT_was_removed_because_it_is_misleading_see_1_20_Release_Notes[-1]
```
instead, which is weird, hmmmpf.

mattip committed:
DEP: deprecate the NPY_SIZEOF_PYARRAYOBJECT macro

seberg committed:
Tune down matti's deprecation to write the error instead.

seberg commented:
OK, I have tried to keep the macro semi-viable with a deprecation warning with the help of Matti. There are many ways how this can just not work probably.

I have tested this, it still requires digging in by the affected end-user if they use it statically:
```
numpy/core/src/multiarray/_multiarray_tests.c.src:2173:5: error: ISO C90 forbids variable length array ‘ret’ [-Werror=vla]
 2173 |     int ret[NPY_SIZEOF_PYARRAYOBJECT];
```
Although, I suppose if you use C99 and enable VLAs, it will actually work and print an error now in this case. The subclassing case will give that error always.

seberg committed:
Tweak macro, so that clang hopefully doesn't complain.

seberg commented:
@mattip is there an issue on PyPy with `PyErr_WriteUnraisable(NULL)`? Can replace it with `Py_None`, not ideal, but also not a big issue.

seberg committed:
Use None instead of NULL in PyErr_WriteUnraisable, pypy seems to have a bug with it

seberg commented:
Let me know whether you like this macro or prefer just to comment it out. I am happy to flip a coin :).

mattip commented:
I think we should comment it out. The attempts to do something clever seem too complicated.

seberg committed:
Just comment it out...

seberg commented:
Sorry for the delay, done now (currently assuming we will just squash merge, so didn't squash).

mattip commented on numpy/core/src/multiarray/_multiarray_tests.c.src:
```diff
@@ -2133,6 +2167,17 @@ run_intp_converter(PyObject* NPY_UNUSED(self), PyObject *args)
     return tup;
 }
 
+static PyObject *
+test_macro(PyObject* NPY_UNUSED(self), PyObject *args)
+{
+    int ret = NPY_SIZEOF_PYARRAYOBJECT;
+    if (ret < 0) {
+        return NULL;
+    }
+    return PyLong_FromLong(ret);
+}
+
+
```
```suggestion

```

mattip commented on numpy/core/src/multiarray/_multiarray_tests.c.src:
```diff
@@ -2311,6 +2359,9 @@ static PyMethodDef Multiarray_TestsMethods[] = {
     {"run_intp_converter",
         run_intp_converter,
         METH_VARARGS, NULL},
+    {"test_macro",
+        test_macro,
+        METH_VARARGS, NULL},
```
```suggestion
    {"test_macro",
        test_macro,
        METH_VARARGS, NULL},
```

seberg replied:
woops :(.

mattip reviewed (commented):
The test is no longer valid and breaks compilation

seberg committed:
Apply suggestions from code review

Co-authored-by: Matti Picus <matti.picus@gmail.com>

eric-wieser commented:
I had a quick look at some CPython types, and concluded that they were also not ABI-safe to extend in C - so I think we're probably justified in breaking this, as long as communicate it well.

charris commented:
Thanks Sebastian. Hadn't noticed that it was updated.

ChristophWWagner commented:
> This speeds up array deallocation and buffer exports, since it
> removes the need to global dictionary lookups. It also somewhat
> simplifies the logic. The main advantage is prossibly less the
> speedup itself (which is not large compared to most things that
> happen in the livetime of an array), but rather that no unnecessary
> work is done for shortlived arrays, which never export a buffer.
> 
> The downside of this approach is that the ABI changes for anyone
> who would be subclassing ndarray in C.
> 
> I am not super invested in this, but I do feel its worth considering. This requires breaking the ABI slightly (so would need release notes), because it extends the array struct (and void, but that is much less likely to be used). Marking as draft since there are probably some smaller moving parts. But since I was looking at the buffer interface today, thought I would wrap this up and show it at least.

I stumbled across this discussion after investigating on an issue that popped up in my package (and obviously many others, too) that low-level interface to numpy from cython.

Currently I (and many others, judging from various issues on gh) do not know how to resolve the distribution mayhem your change caused to a lot of packages. So far I relied on the convention that within one major release no API/ABI breaking shall be rolled out. Now this. 

Please help us on how we, as package developer/maintainer, should deal with the consequences that your breaking changes to the ABI caused. So far I see no good way or workaround to solve that problem and would be happy to be provided with a suggestion on how to deal with this properly.
Should we hard-exclude all numpy versions from 1.20 onwards to maintain compatibility to the existing ecosystem?
Or should we decide to discard all ecosystem behind us and only look on the road ahead? How long until the next change then?

Thank you for your fix to this
Christoph

mattip commented:
Indeed, if a package is using cython and building with 1.20+, the binary will fail when run with pre-1.20 numpy. This is why scipy is very careful to have an "oldest supported numpy" and use it to build with.

eric-wieser commented:
Is there actually an incompatibility here? Or is this cython emitting warnings for something that doesn't need to be warned about?

seberg commented:
Yes, as far as I am aware, it is not possible to actually subclass in cython (if that turns out wrong, we may have a problem).  The problem indeed should "only" be cython warnings, but they may well be turned to errors.

I would assume that things just work if you compile with a new cython and against the oldest NumPy version you support (e.g. 1.16.x).  In that scenairo, the following line in the `__init__.pxd` prevents any warnings/errors IIRC:
```
ctypedef class numpy.ndarray [object PyArrayObject, check_size ignore]
```

@ChristophWWagner can you say which packages broke, and whether they are indeed build with an older NumPy version?  If the problem is building with a new NumPy version and running on old, I would rather see if we can't help you a bit fixing the issue. The ABI was always designed to be "forward compatible" (i.e. compile on old, run on new), never to be "backward compatible".
If existing releases build with old NumPy succumb to this issue, that is a completely different story.

We could also delay this change, meaning backing it out in at least for 1.20.2+... Although, it actually is probably especially advantageous to these Cython modules, due to a reduced conversion overhead from ndarray to memoryview.

seberg commented:
If there is a "simple" (albeit annoying, I am assuming that this annoyance is of an acceptable level), solution, in that all that is needed is compiling with an older NumPy version.  I guess we should add the full error message and explanation to the release notes, more small packages are bound to run into it in the next months when they decide to release new versions/wheels and never had to deal with such version issues.

mattip commented:
We have already been in this situation before. We modified the `PyUFuncObject` in #11175 as part of the 1.16.0 release. I wonder how we communicated then that projects should build with the oldest version of numpy they wish to support?

As for the answer to @eric-wieser's question of is there actually a problem with mixing version, first a digression:

Not all of the numpy C-API goes through functions, some of the capabilities are macros in the headers. When compiling (for instance, with cython), these macros are baked into the c-extension module. 
Cython is very careful here. In [cython 0.29](https://cython.readthedocs.io/en/latest/src/changes.html#id64) (the last item in the release note) they added a `check-size` flag that will allow backward compatible changes: when the runtime returns a struct that is larger than what was expected from the headers, cython will allow it. The reverse situation still raises an error: when the runtime returns a struct that is smaller that what was expected from the headers it is an error.

Given all that, the question becomes "is there a danger that some macro may reach into a field that does not exist in an older version"? The field in question for 1.20 is `_buffer_info`, which appears only in the struct declaration and nowhere else in the header files, so indeed I think it would be safe to use the new headers with an older runtime. We could propose extend the cython enhancement to allow a two-sided safety declaration.
 Although I think there is something to be said for more clearly messaging best practices.

seberg commented:
Relaxing the forward compatibility seems a bit hard. What if we add the pointer, e.g. for the allocator, we probably want functions/macros that use it. As soon as those make it into the Cython API, we lose backward compatibility unless we create extremely careful version checking macros.
In general, maintaining backward compatibility seems, quite frankly, impossible. Since it would not allow us to add new functions and expose them to Cython, unless every single one is actually a macro with some version guard?

I am wondering if we could pre-empt or customize the Cython warning?  Or maybe just make the cython warnings better when `check_size ignore ` is given saying something like: ```numpy.ndarray size changed...; This struct was indicated to grow with newer versions (forward compatibility). When packaging it is typically necessary to compile against the oldest support version, in this case of: numpy.```

---

About `ufunc`, I guess very few Cython libs will actually import/use `ufunc`, so that change – while technically identical – would not probably just go unnoticed?

I am still unsure how much pain we are causing in practice, though.  And there this is only worth so much pain...

mattip commented:
> As soon as those make it into the Cython API, we lose backward compatibility unless we create extremely careful version checking macros.

Only if they are macros, functions call into the runtime so they are fine.

> [I]t would not allow us to add new functions and expose them to Cython ...

We now ship our own version of `numpy/__init__.pxd`, (and `numpy/__init__.cython-30.pxd`) and cython will prefer these to the ones it ships.

seberg commented:
@mattip right, but we cannot add new functions/symbols.  Modifying out `.pxd` is good enough, I am wondering a bit about `import_array()` should that not warn about modified API? Should we have bumped the API version in 1.20 for this, but forgot because it is in a sense not an API change?

ChristophWWagner commented:
> Yes, as far as I am aware, it is not possible to actually subclass in cython (if that turns out wrong, we may have a problem). The problem indeed should "only" be cython warnings, but they may well be turned to errors.
> 
> I would assume that things just work if you compile with a new cython and against the oldest NumPy version you support (e.g. 1.16.x). In that scenairo, the following line in the `__init__.pxd` prevents any warnings/errors IIRC:
> 
> ```
> ctypedef class numpy.ndarray [object PyArrayObject, check_size ignore]
> ```
> 
> @ChristophWWagner can you say which packages broke, and whether they are indeed build with an older NumPy version? If the problem is building with a new NumPy version and running on old, I would rather see if we can't help you a bit fixing the issue. The ABI was always designed to be "forward compatible" (i.e. compile on old, run on new), never to be "backward compatible".
> If existing releases build with old NumPy succumb to this issue, that is a completely different story.
> 
> We could also delay this change, meaning backing it out in at least for 1.20.2+... Although, it actually is probably especially advantageous to these Cython modules, due to a reduced conversion overhead from ndarray to memoryview.

Thank you all for investigating in on this.

The package is `fastmat`, the issue I referred to is this:
https://github.com/EMS-TU-Ilmenau/fastmat/issues/98

When investigating what goes on, I stumbled across these:
https://github.com/scikit-learn-contrib/hdbscan/issues/457
https://github.com/tensorflow/models/issues/9749 (related: https://stackoverflow.com/questions/66373169/tensorflow-2-object-detection-api-numpy-version-errors)
https://ymd_h.gitlab.io/ymd_blog/posts/numpy_1_20_abi_change/

... to finally land here through:
https://numpy.org/doc/stable/release/1.20.0-notes.html#size-of-np-ndarray-and-np-void-changed

The issues appear all to be centered around the step from 1.19 -> 1.20, which is consistent with the change. The last link to a community blog puts out the advice to compile packages with a numpy <= 1.19.5 to get around this problem.

For fastmat, we do the wheel building on Travis-CI with cibuildwheel.
During builds, these requirements are installed:
```  
cython>=0.29
numpy>=1.7
scipy>=1.0
```
(related: https://github.com/EMS-TU-Ilmenau/fastmat/blob/master/requirements-dev.txt)

Here you find the build log from the actual wheel build run:
https://www.travis-ci.org/github/EMS-TU-Ilmenau/fastmat/jobs/760147713

The relevant section for the fastmat issue 98 begins at line 749:
```
Building cp38-manylinux_x86_64 wheel
CPython 3.8 manylinux x86_64
Setting up build environment...
    + /opt/python/cp38-cp38/bin/python -c 'import sys, json, os; json.dump(os.environ.copy(), sys.stdout)'
    + which python
    + which pip
                                                              ✓ 0.04s
Running before_build...
    + sh -c 'python -m pip install -r requirements-dev.txt'
WARNING: You are using pip version 21.0; however, version 21.0.1 is available.
You should consider upgrading via the '/opt/python/cp38-cp38/bin/python -m pip install --upgrade pip' command.
Collecting cython>=0.29
  Downloading Cython-0.29.22-cp38-cp38-manylinux1_x86_64.whl (1.9 MB)
Collecting numpy>=1.7
  Downloading numpy-1.20.1-cp38-cp38-manylinux2010_x86_64.whl (15.4 MB)
Collecting scipy>=1.0
  Downloading scipy-1.6.1-cp38-cp38-manylinux1_x86_64.whl (27.3 MB)
Installing collected packages: numpy, scipy, cython
Successfully installed cython-0.29.22 numpy-1.20.1 scipy-1.6.1
```
During the build no apparent warnings were output.


From what I see the package was built using the latest cython and numpy and fails to operate with older numpy versions. Is there maybe something that I may introduce to the build process that could fix this such that wheels keep on working for all recent versions of numpy?

Thanks!

ChristophWWagner commented:
Just checking the py3.8 CI tests for fastmat (https://www.travis-ci.org/github/EMS-TU-Ilmenau/fastmat/jobs/760152018) and it appears that it there uses the 1.19.5 version as it already comes with the CI environment. For building cython 0.29.22 is used, just as in the cibuildwheel runs during the deployment stage (except there numpy 1.20.1 is used for actually building the wheels, see the log from the previous post).
Note that in the run log here (with numpy 1.19.5) no deployment is done.

During the build the following (known) types of warnings are cast:
```
warning: fastmat/Matrix.pxd:26:10: cpdef variables will not be supported in Cython 3; currently they are no different from cdef variables

/home/travis/virtualenv/python3.8.7/lib/python3.8/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
```

so far nothing unusual in the ci test logs.

seberg commented:
Well, the build is successful and fine, but not compatible with older NumPy versions...  It would be great to give a better error in those cases though (something that makes the solution very obvious).

I think one thing that might make things convenient for you, is to use the scipy meta-package: https://github.com/scipy/oldest-supported-numpy which explicitly picks the smallest NumPy version (I think for a given support python version).  If you build with that, you should be completely fine.

I am not very versed with building, so am a bit unsure how much hassle all of this ends up being...

rgommers commented:
> During builds, these requirements are installed: `numpy>=1.7`

Yes, this is definitely wrong and will cause the kind of problems you are observing. @seberg's suggestion of using `oldest-supported-numpy` is right. Here are the docs that explain in detail: https://numpy.org/devdocs/user/depending_on_numpy.html

rgommers commented:
> It would be great to give a better error in those cases though (something that makes the solution very obvious).

There's a clear error when you do this I believe (at least if the ABI version is incremented (or API, I forgot) - maybe that's the part we missing for 1.20.0?

seberg commented:
The ABI version causes a hard error on any mismatch and would cause forward compatibility to be broken as well as backward compatibility. I.e. it would definitely require a "2.0" release and major churn downstream.

There is this chunk of code in `_import_array`:
```
  if (NPY_FEATURE_VERSION > PyArray_GetNDArrayCFeatureVersion()) {
      PyErr_Format(PyExc_RuntimeError, "module compiled against "\
             "API version 0x%%x but this version of numpy is 0x%%x", \
             (int) NPY_FEATURE_VERSION, (int) PyArray_GetNDArrayCFeatureVersion());
      return -1;
  }
```
But, I just checked and Chuck _did_ increment the API version in 1.20 (which is the same as the `NPY_FEATURE_VERSION`), so I am a bit confused why our error is not raised. Does `cython` manage to check this based on the Python side `import numpy`  before it actually calls the C-side `_import_array()`?

ChristophWWagner commented:
> Well, the build is successful and fine, but not compatible with older NumPy versions... It would be great to give a better error in those cases though (something that makes the solution very obvious).
> 
> I think one thing that might make things convenient for you, is to use the scipy meta-package: https://github.com/scipy/oldest-supported-numpy which explicitly picks the smallest NumPy version (I think for a given support python version). If you build with that, you should be completely fine.
> 
> I am not very versed with building, so am a bit unsure how much hassle all of this ends up being...

Thank you for the hint to `oldest-supported-numpy`. Unfortunately there are two issues with it:
 - It will just break the ci test builds for python 3.4, python 3.6 and python 2.7 for no apparent reason
 - snyk issues vulnerability alerts for numpy 1.14.5 and 1.14.6

Manually setting to numpy 1.14.6 instead breaks for newer python versions (line 719):
https://www.travis-ci.org/github/EMS-TU-Ilmenau/fastmat/jobs/761330130

I feel like in DLL hell...

mattip commented:
> snyk issues vulnerability alerts for numpy 1.14.5 and 1.14.6

But this is the *build* we are talking about, not the *runtime*. The "vulnerability" (which is debatable) is for runtime only.

> It will just break the ci test builds for python 3.4, python 3.6 and python 2.7 

Yes, oldest-supported-numpy is for python3.5 and up, although it should not break 3.6. Are you sure you need to be creating new wheels for python 2.7, 3.4, 3.5? You see users demanding new releases of your package for those versions?

mattip commented:
... if so, please open an issue in oldest-supported-numpy so it can support the python ecosystem better.

rgommers commented:
> * 
@ChristophWWagner you can use different versions to build against for different Python versions, to solve your py34 issue. See the `setup.cfg` of `oldest-supported-numpy`.

> * snyk issues vulnerability alerts for numpy 1.14.5 and 1.14.6 (https://snyk.io/vuln/SNYK-PYTHON-NUMPY-73513)

FYI, that alert was more or less nonsense, _and_ isn't relevant for a build-time issue.

ChristophWWagner commented:
Thank you all for your hep and comments on how to resolve this issue. I have now introduced version-specific minimal requirements and this seems to resove the issue.

Of course you are right that run-time security issues do not play a role here, I have reverted the previous wrong assessment in my setup. Thanks for pointing this out!

`oldest-supported-numpy` caused a version to be selected that was not supported by another package, which led to a length dependency resolution process in pip. To reduce CI build load I decided to specify the oldest combinations for which wheels exist. This should do the job.

Unfortunately we have some mainframe systems running that still run python2 code. That's why we still keep the wheels in.

As some of you were asking on how to communicate this issue better I would suggest to reqork the error message. First time I have seen this I thought of a bug within the package and (as a user) I thought there was nothing that **I could do** here. Only after seeing this message repeat as a pattern I started investigating and found out that there is a way of arriving at an inconsistent dependency state and could start to resolve that. Maybe pointing to this idea could already help communicating the underlying issue.

Again, thanks for all your help!

ChristophWWagner commented:
> @ChristophWWagner you can use different versions to build against for different Python versions, to solve your py34 issue. See the `setup.cfg` of `oldest-supported-numpy`.
> 
> > * snyk issues vulnerability alerts for numpy 1.14.5 and 1.14.6 (https://snyk.io/vuln/SNYK-PYTHON-NUMPY-73513)
> 
> FYI, that alert was more or less nonsense, _and_ isn't relevant for a build-time issue.

out of curiosity and aside from the buid-time issue: why is this security alert bogus?

eric-wieser commented:
> out of curiosity and aside from the buid-time issue: why is this security alert bogus?

Because it's equivalent to saying "Python has a vulnerability because `pickle.load` allows arbitrary code execution".
The true statement is that "code that uses `np.load` unsafely is vulnerable", not "`np.load` is unsafe".

This comment does a good job of conveying that: https://github.com/numpy/numpy/issues/12759#issuecomment-459801406

rgommers commented:
This is still causing problems:
- https://github.com/scipy/scipy/issues/14652
- https://stackoverflow.com/questions/66060487/valueerror-numpy-ndarray-size-changed-may-indicate-binary-incompatibility-exp

That's a pretty popular SO issue. Looking at the comments, we knew this was quite a risky change - looks like we should learn this lesson and not do this kind of thing again unless we bump the major version.

winash12 commented:
I have only one version of numpy on my machine and I still get this error.

mattip commented:
> looks like we should learn this lesson and not do this kind of thing again unless we bump the major version.

I am not convinced by that issue. If I read it correctly, the user was building locally using `python setup.py install`, wit a newer NumPy, but then ran their code with an older NumPy. Building locally with `python setup.py` avoids all the attempts by a build system to pin required package versions.

The alternative is to never change NumPy structs used by cython, which is also painful.

Just to expand a bit: the problem is due to
- the user having a version of numpy in their environment before this change (say 1.19)
- installing a package that somehow builds using cython and a newer version of NumPy (say 1.20)

Then cython codes the size of the new (larger) struct, and raises an error when seeing the smaller one. The other direction (building with 1.19, running with 1.21) works, although it may warn.

rgommers commented:
> The alternative is to never change NumPy structs used by cython, which is also painful.

We can add things, just not remove them, right? And we actually do that semi-regularly? 

Here's another example: https://github.com/RaRe-Technologies/gensim/issues/3085#issuecomment-804710317. It's really hard to figure out what is actually caused by this PR and what may be user error or error of another project building wheels a certain way. But the bug reports seem compatible with the actual impact we expect (best summarized with @eric-wieser's comment [here](https://github.com/numpy/numpy/pull/16938#issuecomment-731618597)). And given this was a "it's not that important, but let's see what y'all think" kind of change anyway, my impression is that we shouldn't have merged it.

mattip commented:
Looking at that example, once they figured out the problem, there is this message:
> Using the `oldest-supported-numpy` as build dependency worked, and the wheels got built on the first try. I don't think that's ever happened before ;)

so it seems like this PR has the side-effect of helping projects clean up their builds. :)

We never remove fields. And if we reject the approach in this PR neither will we be able to add fields. We do not do that regularly, this was the first attempt to change `PyArrayObject` in many years. 

> my impression is that we shouldn't have merged it

We wanted to see what would happen if we changed the size: this is a canary for the further proposed changes to structs that appear in the cython API (`PyArrayObject` and `PyArray_Descr` are the biggest but there are more). If we cannot change structs then we cannot accept NEP 49 (allocator strategies) and the dtype refactor will have to be much more limited. We did discuss this many times at the developer and triage meetings before we decided to go ahead with it.

rgommers commented:
> We wanted to see what would happen if we changed the size: this is a canary for the further proposed changes

My conclusion is that each time we do something like this (the C API change in 1.19.2 comes to mind as well), we get multiple projects linking to the PR in question claiming we broke something - which we did, it's just not clear that that is actually the cause of their problem or not. Doing more of this seems unhealthy, at least without better diagnostics / safeguards.

> But, I just checked and Chuck _did_ increment the API version in 1.20 (which is the same as the `NPY_FEATURE_VERSION`), so I am a bit confused why our error is not raised. Does `cython` manage to check this based on the Python side `import numpy` before it actually calls the C-side `_import_array()`?

This one (from @seberg's [comment](https://github.com/numpy/numpy/pull/16938#issuecomment-789987915) above) we should get to the bottom of. If the problem is indeed just "using a package that was built against a newer numpy with an older numpy at runtime", then our error should say that on import. If we continue to get:
```
ValueError: numpy.ndarray size changed, may indicate binary incompatibility
```
from Cython, that'll be super unhelpful for everyone.

winash12 commented:
> > looks like we should learn this lesson and not do this kind of thing again unless we bump the major version.
> 
> I am not convinced by that issue. If I read it correctly, the user was building locally using `python setup.py install`, wit a newer NumPy, but then ran their code with an older NumPy. Building locally with `python setup.py` avoids all the attempts by a build system to pin required package versions.
> 
> The alternative is to never change NumPy structs used by cython, which is also painful.
> 
> Just to expand a bit: the problem is due to
> 
> * the user having a version of numpy in their environment before this change (say 1.19)
> * installing a package that somehow builds using cython and a newer version of NumPy (say 1.20)
> 
> Then cython codes the size of the new (larger) struct, and raises an error when seeing the smaller one. The other direction (building with 1.19, running with 1.21) works, although it may warn.

Just to clarify. I only have one version of numpy on my machine that I installed myself using setup.py. The problem here is h5py which uses a numpy egg file while building that is version 1.17. This new egg file is present within the h5py installation. That is the reason for the error on my machine.  Please this comment - https://github.com/h5py/h5py/issues/1955#issuecomment-908835052

mattip commented:
@winash12 Since that issue is closed I don't want to rehash it here. The question here is "can we give a better error when a package is built against a newer numpy but uses an older numpy at runtime".

seberg commented:
Do I read the latest discussion right if I say:  Real progress here would be trying to make the cython error more clear, or even seeing if we can get cython to call `import_array()` first (which probably already gives a better warning/error?)

rgommers commented:
Yes, that sounds right. If we could just make it say "Installed numpy version is older than the numpy version that this package was built against" or something like that, that would make it a lot less painful to debug issues that are potentially due to PRs like this one.
